peerstore.go

```diff
@@ -1,254 +1,41 @@
-
-	// UpdateAddrs updates the addresses associated with the given peer that have
-	// the given oldTTL to have the given newTTL.
-	UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL time.Duration)
-
-	// Addresses returns all known (and valid) addresses for a given peer
-	Addrs(p peer.ID) []ma.Multiaddr
-
-	// AddrStream returns a channel that gets all addresses for a given
-	// peer sent on it. If new addresses are added after the call is made
-	// they will be sent along through the channel as well.
-	AddrStream(context.Context, peer.ID) <-chan ma.Multiaddr
-
-	// ClearAddresses removes all previously stored addresses
-	ClearAddrs(p peer.ID)
-
-	// Peers returns all of the peer IDs stored in the AddrBook
-	Peers() []peer.ID
-}
-
-// KeyBook tracks the Public keys of Peers.
-type KeyBook interface {
-	PubKey(peer.ID) ic.PubKey
-	AddPubKey(peer.ID, ic.PubKey) error
-
-	PrivKey(peer.ID) ic.PrivKey
-	AddPrivKey(peer.ID, ic.PrivKey) error
-}
-
-type keybook struct {
-	pks map[peer.ID]ic.PubKey
-	sks map[peer.ID]ic.PrivKey
-
-	sync.RWMutex // same lock. wont happen a ton.
-}
-
-func newKeybook() *keybook {
-	return &keybook{
-		pks: map[peer.ID]ic.PubKey{},
-		sks: map[peer.ID]ic.PrivKey{},
-	}
-}
-
-func (kb *keybook) Peers() []peer.ID {
-	kb.RLock()
-	ps := make([]peer.ID, 0, len(kb.pks)+len(kb.sks))
-	for p := range kb.pks {
-		ps = append(ps, p)
-	}
-	for p := range kb.sks {
-		if _, found := kb.pks[p]; !found {
-			ps = append(ps, p)
-		}
-	}
-	kb.RUnlock()
-	return ps
-}
-
-func (kb *keybook) PubKey(p peer.ID) ic.PubKey {
-	kb.RLock()
-	pk := kb.pks[p]
-	kb.RUnlock()
-	if pk != nil {
-		return pk
-	}
-	pk, err := p.ExtractPublicKey()
-	if err == nil && pk != nil {
-		kb.Lock()
-		kb.pks[p] = pk
-		kb.Unlock()
-	}
-	return pk
-}
-
-func (kb *keybook) AddPubKey(p peer.ID, pk ic.PubKey) error {
-
-	// check it's correct first
-	if !p.MatchesPublicKey(pk) {
-		return errors.New("ID does not match PublicKey")
-	}
-
-	kb.Lock()
-	kb.pks[p] = pk
-	kb.Unlock()
-	return nil
-}
-
-func (kb *keybook) PrivKey(p peer.ID) ic.PrivKey {
-	kb.RLock()
-	sk := kb.sks[p]
-	kb.RUnlock()
-	return sk
-}
-
-func (kb *keybook) AddPrivKey(p peer.ID, sk ic.PrivKey) error {
-
-	if sk == nil {
-		return errors.New("sk is nil (PrivKey)")
-	}
-
-	// check it's correct first
-	if !p.MatchesPrivateKey(sk) {
-		return errors.New("ID does not match PrivateKey")
-	}
-
-	kb.Lock()
-	kb.sks[p] = sk
-	kb.Unlock()
-	return nil
-}
-
-type peerstore struct {
-	*keybook
-	*metrics
+	KeyBook
 	AddrBook
-
-	// store other data, like versions
-	//ds ds.ThreadSafeDatastore
-	// TODO: use a datastore for this
-	ds     map[string]interface{}
-	dslock sync.Mutex
+	PeerMetadata
 
 	// lock for protocol information, separate from datastore lock
 	protolock sync.Mutex
 }
 
 // NewPeerstore creates a threadsafe collection of peers.
-func NewPeerstore() Peerstore {
+func NewPeerstoreWith(kb KeyBook, ab AddrBook, md PeerMetadata) Peerstore {
```

Comment:
👍

Reply:
I would drop the `With` as it is implied by the fact that you are passing those parameters in ;)

Reply:
Also if you do change it make sure to update the godocs to reflect any changes.

Reply:
Sounds good. I'll quickly push that 👍